posts/2021/01/enhancements-to-the-guvectorize-decorator.md

```diff
@@ -0,0 +1,150 @@
+<!--
+.. title: Enhancements to the guvectorize decorator
+.. slug: enhancements-to-the-guvectorize-decorator
+.. date: 2021-01-20 08:00:00 UTC-00:00
+.. author: Guilherme Leobas
+.. tags: Labs, Numba
+.. category:
+.. link:
+.. description:
+.. type: text
+-->
+
+Starting on Numba 0.53, Numba will ship with an enhanced version of the `@guvectorize` decorator. Similar to the [@vectorize](https://numba.pydata.org/numba-doc/dev/user/vectorize.html#the-vectorize-decorator) decorator, [@guvectorize](https://numba.pydata.org/numba-doc/dev/user/vectorize.html#the-guvectorize-decorator) now has two modes of operation: 
+
+- Eager, or decoration-time compilation and
+- Lazy, or call-time compilation
+
+Before, only the eager approach was supported. In this mode, users are required to provide a list of concrete supported types beforehand as its first argument. Now, this list can be omitted if desired and as one calls it, Numba dynamically generates new kernels for previously unsupported types.
+
+<!-- TEASER_END -->
+
+## NumPy Universal Functions
+
+NumPy introduces a type of function called Universal functions or ufuncs. Ufuncs are functions that operate on `ndarrays` element-by-element. [Examples](https://numpy.org/doc/stable/reference/ufuncs.html#available-ufuncs) of universal functions are `np.log` and `np.log2`, which compute the natural and base-2 logarithms, respectively. Alongside ufuncs, NumPy also introduces the notion of generalized ufuncs or gufuncs. While the former is limited to element-by-element operations, the latter supports sub-array by sub-array operations at a time.
+
+Creating new NumPy ufuncs is not an easy process and may require one to [write some C code](https://numpy.org/doc/stable/user/c-info.ufunc-tutorial.html). Numba extends the NumPy mechanism for registering and using (generalized) universal functions with two decorators: `@vectorize` and `@guvectorize`. Those decorators allow one to easily create universal functions from Python, leaving the gruntwork to Numba.
+
+For instance, consider the function `guvec`, which adds a scalar to every element in an array:
```

Comment:
It would be worth mentioning the caveat that `@guvectorize`d functions have to have the return array passed in as an argument.